### ISSUE TITLE
Implement factorial power

### ISSUE THREAD

**Issue by axkr** [axkr/symja_android_library#184]
See:
- https://en.wikipedia.org/wiki/Falling_and_rising_factorials

Implement the FactorialPower function in this prepared function "template" for 2 and 3 arguments:
- https://github.com/axkr/symja_android_library/blob/a08cee32deb0e78c7f12a54e153825f731168233/symja_android_library/matheclipse-core/src/main/java/org/matheclipse/core/builtin/NumberTheory.java#L1952

See:
- http://www.luschny.de/math/factorial/
- http://www.luschny.de/math/factorial/FastFactorialFunctions.htm
- https://www.geeksforgeeks.org/bigintegermath-factorial-function-guava-java/

Discuss here:
- https://gitter.im/symja_android_library/Lobby

**Comment by OleksandrMariupol** [comment]
Hi axkr. Thanks for your time. I have some ideas. I want to clarify a few details. Should I consider cases when x> 0 and when x <0?

**Comment by axkr** [comment]
Please start with the falling factorial as ´FactorialPower(x,n)` with `n>=0`
- https://en.wikipedia.org/wiki/Falling_and_rising_factorials

```
x*(x-1)*(x-2)....(x-n+1)
```

**Comment by OleksandrMariupol** [comment]
Hi axkr. Thank you for your time. I can not find tests for FactorialPower. Could you please tell me, maybe there are some tests? 

Regards

**Comment by axkr** [comment]
Some  JUnit test cases could be:

```

	public void testFactorialPower() {
		check("FactorialPower(3, 2)", //
				"6");
		check("FactorialPower(2, 2)", //
				"2");
		check("FactorialPower(1, 2)", //
				"0");
		check("FactorialPower(0, 2)", //
				"0");
		check("FactorialPower(-1, 2)", //
				"2");
		check("FactorialPower(-2, 2)", //
				"6");
		check("FactorialPower(-3, 2)", //
				"12");
		check("FactorialPower(100, 3)", //
				"970200");
		check("FactorialPower(-100, 3)", //
				"-1030200");
	}
	
```

**Comment by OleksandrMariupol** [comment]
Hi axkr. Thank you for your answer. I have created the pull request https://github.com/axkr/symja_android_library/pull/194 Could you please see it. And let me know, if it needs, what should be changed. 
Regards

**Comment by axkr** [comment]
Thanks for the PR. I just merged your PR.

Some further suggestions you can try to improve the evaluation:
- can you mention the issue as `#184` in your commits, so that it shows up in the history of this issue.
- with `ast.arg2().toIntDefault(Integer.MIN_VALUE)` you can check if its an ìnt` and optimize the while loops.
- with `expr.isZero()` you can test if an expression is `0`
- can you add some JUnit tests for FactorialPower() with 3 arguments

**Comment by axkr** [comment]
Could you please check if your implementation also works for complex and fraction numbers.

Example: 

```
check("FactorialPower(2-I, 2)", //
				"1-I*3");
```

**Comment by OleksandrMariupol** [comment]
Hi axkr. Thank you for your answer and your time .Can you please clarify some details. As I know, for fraction numbers, we can use Euler's Integral Formula "Gamma Function" to find factorial of 1/2 or 3.1. For example if it needs to find (1/2)! for n = 4. If we do not setup h"our step", what step will be between x1 and x2 in case of fraction numbers" (1/2)! = 1/2 * (1/2 - 1) * (1/2 - 2) * ... , "I think it can be one" or something else"?  If we setup h "our step", for example h = 1/2 it will be "(1/2)! = 1/2 * (1/2 - 1/2) * (1/2 - 1) *.... "?
Regards

**Comment by axkr** [comment]
Do you mean the Gamma formulas for real values?
- https://en.wikipedia.org/wiki/Falling_and_rising_factorials#Properties

Do you mean?
```
x*(x-1*h)*(x-2*h)....(x-(n-1)*h)
```

**Comment by OleksandrMariupol** [comment]
Hi axkr.  This video (0.5)! https://www.youtube.com/watch?v=dBmApl6gkII and we can calculate with this formula for example (6.5)! and yes we can use it in https://en.wikipedia.org/wiki/Falling_and_rising_factorials#Properties. Yes, I mean x*(x-1*h)*(x-2*h)....(x-(n-1)*h).   I just want to check myself again, what I need to do".  My implementation also works for fraction numbers, test it today, now I am thinking about  complex numbers.

**Comment by OleksandrMariupol** [comment]
Hi axkr. Thank you for your answer. I have created the pull request https://github.com/axkr/symja_android_library/pull/195 Could you please see it. And let me know, what should be changed or add.
Regards

**Comment by OleksandrMariupol** [comment]
Hi axkr. Thank you for your time. I have created the pull request https://github.com/axkr/symja_android_library/pull/196 Could you please see it. I have changed  some logic. Now it is better. Let me know, if something.
Regards

**Comment by OleksandrMariupol** [comment]
Hi. Now, we can calculate falling factorial. When h>0: f(x) * f(x -h) * f(x-2h) ...... * f(x - (n-1)h) -> formula for falling factorial. I think,  when h < 0, this formula is equivalent to, f(x) * f(x +h) * f(x+2h) ...... * f(x + (n-1)h) ->  formula for rising factorial. My idea is, that, if user enter h = -1 or h = -0.5 etc h < 0 , algorithm should calculate rising factorial,  change algorithm logic in method FactorialPower for this purpose. What you think about this?

**Comment by axkr** [comment]
The current implementation will be released in 2.0.0. Please open new issue if necessary.
